python/gui/auto_generated/editorwidgets/core/qgswidgetwrapper.sip.in

```diff
@@ -78,7 +78,7 @@ Access the widget managed by this wrapper
 %End
 
 
-    void setConfig( const QVariantMap &config );
+    virtual void setConfig( const QVariantMap &config );
```

Comment:
Do we need to make this virtual? All other subclasses will inspect the config on first use

Reply:
In short: I would prefer to have all widgets support the same coding pattern and therefore avoid having initialization of configuration-dependent members in different ways for consistency and maintainability.

If there is a compelling reason to do the initialization on setter-time, I'd prefer to have a private virtual member with default empty body to implements this behaviour in order to avoid bypassing the main setConfig method.

Reply:
I don't understand you comment.
This is overridden because after having tried different approaches I think it's best to resolve the layer when config is set, because the precondition in other calls (see: `QgsValueRelationFieldFormatter::createCache`) is that "Layer" has been already resolved.

Reply:
Sorry, you second comment did not appear when I was writing my reply.
I think it's a good idea to add a private virtual method (checkConfig or resolveConfig might be a good name but it's private so it doesn't really matter).

Reply:
> see: QgsValueRelationFieldFormatter::createCache) is that "Layer" has been already resolved.>

Isn't the field formatter independent from the widget? It takes a config, layer and field index, but no widget wrapper.

Reply:
You said it: it takes a config ans it's called by the widget. If the config wasn't already cleaned we would need to re-apply the same reconcile logic here (and in other places), IMO is better to fix the config as soon as we can.

Reply:
Hmm... I think there is no guarantee that the config has been pre-processed by a widgetwrapper or the config before. This class was factored out of the widget to be used by independent code (e.g. server where there are no widgets). So I think the same logic needs to be reapplied here.

Reply:
ok, then I'll take a different approach and implement a static getLayer(config) somewhere to be re-used any time we need to get a layer reference in the dialog, the wrapper or the formatter.

Reply:
That sounds perfect :+1: